// @grpc/grpc-js-xds/build/src/generated/envoy/config/core/v3/HealthStatus.js
"use strict";
// Original file: deps/envoy-api/envoy/config/core/v3/health_check.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.HealthStatus = void 0;
/**
 * Endpoint health status.
 */
var HealthStatus;
(function (HealthStatus) {
    /**
     * The health status is not known. This is interpreted by Envoy as *HEALTHY*.
     */
    HealthStatus[HealthStatus["UNKNOWN"] = 0] = "UNKNOWN";
    /**
     * Healthy.
     */
    HealthStatus[HealthStatus["HEALTHY"] = 1] = "HEALTHY";
    /**
     * Unhealthy.
     */
    HealthStatus[HealthStatus["UNHEALTHY"] = 2] = "UNHEALTHY";
    /**
     * Connection draining in progress. E.g.,
     * `<https://aws.amazon.com/blogs/aws/elb-connection-draining-remove-instances-from-service-with-care/>`_
     * or
     * `<https://cloud.google.com/compute/docs/load-balancing/enabling-connection-draining>`_.
     * This is interpreted by Envoy as *UNHEALTHY*.
     */
    HealthStatus[HealthStatus["DRAINING"] = 3] = "DRAINING";
    /**
     * Health check timed out. This is part of HDS and is interpreted by Envoy as
     * *UNHEALTHY*.
     */
    HealthStatus[HealthStatus["TIMEOUT"] = 4] = "TIMEOUT";
    /**
     * Degraded.
     */
    HealthStatus[HealthStatus["DEGRADED"] = 5] = "DEGRADED";
})(HealthStatus = exports.HealthStatus || (exports.HealthStatus = {}));
